apps/services/user-notification/src/app/modules/notifications/notifications.controller.ts

```diff
@@ -1,81 +1,138 @@
-import { Inject } from '@nestjs/common'
 import {
-  Controller,
-  Post,
+  Inject,
+  Body,
+  Get,
+  Param,
+  Query,
+  CacheInterceptor,
+  UseInterceptors,
   BadRequestException,
-  Req,
-  HttpCode,
 } from '@nestjs/common'
+import { Controller, Post, HttpCode } from '@nestjs/common'
 import {
   ApiOkResponse,
   ApiBody,
   ApiExtraModels,
   getSchemaPath,
+  ApiOperation,
 } from '@nestjs/swagger'
-import { validate, ValidationError } from 'class-validator'
-import { Request } from 'express'
 import type { Logger } from '@island.is/logging'
 import { LOGGER_PROVIDER } from '@island.is/logging'
-import {
-  NewDocumentMessage,
-  Message,
-  TypeValidator,
-  ValidatorTypeMap,
-} from './dto/createNotification.dto'
+import { CreateNotificationDto } from './dto/createNotification.dto'
 import { InjectQueue, QueueService } from '@island.is/message-queue'
 import { CreateNotificationResponse } from './dto/createNotification.response'
 
-const throwIfError = (errors: ValidationError[]): void => {
-  if (errors.length > 0) {
-    throw new BadRequestException(
-      errors.reduce(
-        (acc, e) => acc.concat(Object.values(e.constraints ?? {})),
-        [] as string[],
-      ),
-    )
-  }
-}
-
-// Validates the message POST parameter. The top-level message object has a
-// `type` attribute that defines how the rest of the object is validated.
-// class-validator doesn't seem to support that and this is the best solution
-// I could find.
-const validateMessage = async (body: Request['body']): Promise<Message> => {
-  const type = new TypeValidator()
-  Object.assign(type, body)
-  throwIfError(await validate(type, { forbidUnknownValues: false }))
+import { CreateHnippNotificationDto } from './dto/createHnippNotification.dto'
+import { Documentation } from '@island.is/nest/swagger'
+import { HnippTemplate } from './dto/hnippTemplate.response'
 
-  const ValidatorClass = ValidatorTypeMap[type.type]
-  const message = new ValidatorClass()
-  Object.assign(message, body)
-  throwIfError(await validate(message, { forbidUnknownValues: true }))
-
-  return message
-}
+import { NotificationsService } from './notifications.service'
 
 @Controller('notifications')
-@ApiExtraModels(NewDocumentMessage)
+@ApiExtraModels(CreateNotificationDto)
+@UseInterceptors(CacheInterceptor) // auto-caching GET responses
 export class NotificationsController {
```

Comment:
Should split this up into multiple versions and REST-ful controllers. E.g.

```
POST /v1/notifications
POST /notifications  - v1 for backwards compatibility.

POST /v2/notifications - new endpoint
GET /v2/templates
GET /v2/templates/:templateId
```

Check out the [docs](https://docs.nestjs.com/techniques/versioning). Can enable versioning in your bootstrap file and then configure version for each controller. Good to also split them up in multiple files and even folders (v1 vs v2).